File: plugins/tables.lua
-- NESBuilder plugin
-- tables.lua

local plugin = {
    author = "SpiderDave",
    default = false,
}

function plugin.onInit()
    local buttonWidth = config.buttonWidth*7.5
    local buttonHeight = 27
    local inputWidth = buttonWidth*2
    local inputHeight = 29
    local control
    local pad=6
    local left = pad * 1.5
    local top = pad * 1.5
    local x,y = left, top
    
    local colX = 170
    
    --NESBuilder:makeTabQt{name="tabTableTools", text="Table Tools"}
    makeTab{name="tabTableTools", text="Table Tools"}
    NESBuilder:setTabQt("tabTableTools")
    
    control=NESBuilder:makeLabelQt{x=x,y=y,w=buttonWidth, clear=true,text="Sine Table Generator"}
    control.setFont("Verdana", 13)
    y = y + control.height + pad * 2
    
    control=NESBuilder:makeLabelQt{x=x,y=y,w=buttonWidth, clear=true,text="Label:"}
    control.setFont("Verdana", 12)
    push(x)
    x = colX
    control = NESBuilder:makeLineEdit{x=x,y=y,w=inputWidth*.5,h=inputHeight, name="sineLabel", text='sineTable'}
    y = y + control.height + pad
    
    x = pop()
    control=NESBuilder:makeLabelQt{x=x,y=y,w=buttonWidth, clear=true,text="Points per quarter:"}
    control.setFont("Verdana", 12)
    push(x)
    x = colX
    control = NESBuilder:makeLineEdit{x=x,y=y,w=32,h=inputHeight, name="sinePoints", text='16'}
    y = y + control.height + pad
    
    x = pop()
    control=NESBuilder:makeLabelQt{x=x,y=y,w=buttonWidth, clear=true,text="Amplitude: "}
    control.setFont("Verdana", 12)
    push(x)
    x = colX
    control = NESBuilder:makeLineEdit{x=x,y=y,w=32,h=inputHeight, name="sineAmplitude", text='100'}
    y = y + control.height + pad
    
    control = NESBuilder:makeCheckbox{x=x,y=y,name="sine16Bit", text="Use 16-bit", value=0}
    y = y + control.height + pad
    
    x = pop()
    control = NESBuilder:makeButton{x=x,y=y,w=100,h=buttonHeight,name="buttonSineTable",text="Generate"}
    push(x)
    x = x + control.width + pad
    
    control = NESBuilder:makeButton{x=x,y=y,w=100,h=buttonHeight,name="buttonSineTableClear",text="Clear"}
    x=pop()
    y = y + control.height + pad
    
    control = NESBuilder:makeTextEdit{x=x,y=y,w=700,h=450,name="textSineTable"}
    y = y + control.height + pad
end

function plugin.buttonSineTableClear_cmd(t)
    NESBuilder:getControl('textSineTable').clear()
    data.project.sineTables = {}
end

function plugin.onLoadProject()
    data.project.sineTables = data.project.sineTables or {}
    updateSineTables()
end

function plugin.buttonSineTable_cmd(t)
    local control
    local points, amp, label
    
    if not pcall(function()
        label = NESBuilder:getControl('sineLabel').text
        points = int(NESBuilder:getControl('sinePoints').text)
        amp = int(NESBuilder:getControl('sineAmplitude').text)
    end) then
        print('invalid value(s)')
        return
    end
    
    local t = {
        label = label,
        points = points,
        amp = amp,
    }
    table.insert(data.project.sineTables, t)
    
    -- This index will be used to handle duplicate labels
    labelCount = {}
    for i,v in ipairs(data.project.sineTables) do
        labelCount[v.label] = (labelCount[v.label] or 0) + 1
        v.i = labelCount[v.label]
    end
    
    updateSineTables()
end

function updateSineTables()
    local generateSineTable = NESBuilder:importFunction('plugins.tables','generateSineTable')
    local makeTableData = NESBuilder:importFunction('plugins.tables','makeTableData')
    
    local control = NESBuilder:getControl('textSineTable')
    control.clear()
    
    for i,v in ipairs(data.project.sineTables) do
        local t = generateSineTable(v.points, v.amp, 2)
        
        local labelBase = string.format('%s%s', v.label, v.i>1 and v.i or '')
        
        control.print(labelBase..'_low:')
        control.print(string.format('    ; Points per quarter cycle: %d', v.points))
        control.print(string.format('    ; Amplitude: %d', v.amp))
        control.print(makeTableData(t.q1Low))
        control.print()
        control.print(makeTableData(t.q2Low))
        control.print()
        control.print(makeTableData(t.q3Low))
        control.print()
        control.print(makeTableData(t.q4Low))
        control.print()
        control.print(labelBase..'_high:')
        control.print(makeTableData(t.q1High))
        control.print()
        control.print(makeTableData(t.q2High))
        control.print()
        control.print(makeTableData(t.q3High))
        control.print()
        control.print(makeTableData(t.q4High))
        control.print()
    end
end

function plugin.onBuild()
    local filename, out
    
    if #data.project.sineTables == 0 then return end
    
    filename = data.folders.projects..projectFolder.."code/tables.asm"
    out = NESBuilder:getControl('textSineTable').text
    util.writeToFile(filename, 0, out, true)
    
    print("File created: ".. filename)
end

return plugin

File: plugins/tables.py
import math
from math import cos, sin, pi

# points is number of points per quarter cycle
def generateSineTable(points = 20, amplitude = 100, intSize = 1):
    quarter_sine_table = []
    half_sine_table = []
    sine_table = []

    for x in range(0, points):
        n = sin(2 * pi * (90/points * x) / 360)
        n = int(n * amplitude)
        quarter_sine_table.append(n)
    
#    def diffSine(sineTable):
#        newT = []
#        for i, t in enumerate(sineTable):
#            if i > 0:
#                t = t-sineTable[i-1]
#            if t<0:
#                t = t % 256
#            newT.append(t-sineTable[i-1])
#        return newT
    
    if amplitude>256:
        intSize = 2
    
    diffSine = lambda t: [(x-t[i-1]) % (256**intSize) if i>0 else 0 for i, x in enumerate(t)]
    q1 = quarter_sine_table
    q2 = quarter_sine_table[::-1]
    
    q3 = [-x % (256**intSize) for x in q1]
    q4 = [-x % (256**intSize) for x in q2]
    q1Diff=diffSine(q1)
    q2Diff=diffSine(q2)
    q3Diff=diffSine(q3)
    q4Diff=diffSine(q4)
    
    low = lambda t: [x % 256 for x in t]
    high = lambda t: [math.floor(x / 256) for x in t]
    
    
    q1Low = low(q1)
    q1High = high(q1)
    q2Low = low(q2)
    q2High = high(q2)
    q3Low = low(q3)
    q3High = high(q3)
    q4Low = low(q4)
    q4High = high(q4)
    
    return dict(
        q1=q1,
        q2=q2,
        q3=q3,
        q4=q4,
        q1Low=q1Low,
        q1High=q1High,
        q2Low=q2Low,
        q2High=q2High,
        q3Low=q3Low,
        q3High=q3High,
        q4Low=q4Low,
        q4High=q4High,
        full=q1+q2+q3+q4,
        half=q1+q2,
        quarter=q1,
        
        q1Diff=q1Diff,
        q2Diff=q2Diff,
        q3Diff=q3Diff,
        q4Diff=q4Diff,
        fullDiff=q1Diff+q2Diff+q3Diff+q4Diff,
        halfDiff=q1Diff+q2Diff,
        quarterDiff=q1Diff,
        
        intSize = intSize,
    )

def chunker(seq, size):
    res = []
    for el in seq:
        res.append(el)
        if len(res) == size:
            yield res
            res = []
    if res:
        yield res

def makeTableData(data, intSize=1, indent=4, nItems=8):
    # make sure it's a list, tuple, etc
    if type(data) in (int, str):
        data = [data]
    out=''
    
    try:
        if max(data) > 256:
            intSize = 2
    except:
        pass
    
    for chunk in chunker(data, nItems):
        newData = []
        for i, item in enumerate(chunk):
            if type(item) is int:
                newData.append('${0:02x}'.format(item))
            else:
                newData.append(item)
        if intSize > 1:
            out+=' '*indent+'dw '+', '.join(newData)+'\n'
        else:
            out+=' '*indent+'db '+', '.join(newData)+'\n'
    out = out.rstrip()
    return out
